src/libsyntax/codemap.rs

```diff
@@ -579,38 +579,75 @@ impl CodeMap {
 
     /// Returns a new span representing just the end-point of this span
     pub fn end_point(&self, sp: Span) -> Span {
-        let hi = sp.hi().0.checked_sub(1).unwrap_or(sp.hi().0);
-        let hi = self.get_start_of_char_bytepos(BytePos(hi));
-        let lo = cmp::max(hi.0, sp.lo().0);
-        sp.with_lo(BytePos(lo))
+        let pos = sp.hi().0;
+
+        let width = self.find_width_of_character_at_span(sp, false);
+        let corrected_end_position = pos.checked_sub(width).unwrap_or(pos);
+
+        let end_point = BytePos(cmp::max(corrected_end_position, sp.lo().0));
+        sp.with_lo(end_point)
     }
 
     /// Returns a new span representing the next character after the end-point of this span
     pub fn next_point(&self, sp: Span) -> Span {
-        let hi = sp.lo().0.checked_add(1).unwrap_or(sp.lo().0);
-        let hi = self.get_start_of_char_bytepos(BytePos(hi));
-        let lo = cmp::max(sp.hi().0, hi.0);
-        Span::new(BytePos(lo), BytePos(lo), sp.ctxt())
+        let pos = sp.lo().0;
+
+        let width = self.find_width_of_character_at_span(sp, true);
+        let corrected_next_position = pos.checked_add(width).unwrap_or(pos);
+
+        let next_point = BytePos(cmp::max(sp.hi().0, corrected_next_position));
+        Span::new(next_point, next_point, sp.ctxt())
     }
 
-    fn get_start_of_char_bytepos(&self, bpos: BytePos) -> BytePos {
-        let idx = self.lookup_filemap_idx(bpos);
-        let files = self.files.borrow();
-        let map = &(*files)[idx];
+    /// Finds the width of a character, either before or after the provided span.
+    fn find_width_of_character_at_span(&self, sp: Span, forwards: bool) -> u32 {
+        // Disregard malformed spans and assume a one-byte wide character.
+        if sp.lo() > sp.hi() {
+            return 1;
+        }
```

Comment:
Did you trigger this check anywhere? I'd like to keep it, but there have been efforts to avoid creating malformed spans in the first place.

Reply:
I don't create any spans that would trigger this. I know that if I attempt to assert that `lo < hi` then it fails when compiling the compiler, so I think this check is necessary for now.

Reply:
Yes, we should definitely keep the check. Checks like this are peppered throughout the compiler because the compiler is making bad spans. Filed #47504 with my thoughts on the matter.